File: mcspring-api/mcspring-subcommands/src/test/java/in/kyle/mcspring/subcommands/TestTabCompletion.java
package in.kyle.mcspring.subcommands;

import org.bukkit.command.CommandSender;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.function.Consumer;
import java.util.function.Function;

import in.kyle.mcspring.subcommands.PluginCommandBase.State;
import in.kyle.mcspring.subcommands.tab.TabDiscovery;
import in.kyle.mcspring.test.MCSpringTest;
import lombok.var;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.*;

@MCSpringTest
public class TestTabCompletion {
    
    @Autowired
    TabDiscovery tabDiscovery;
    TestSender sender;
    
    @BeforeEach
    void setup() {
        sender = spy(TestSender.class);
    }
    
    @Test
    void testTabWithDirectExecution() {
        class Test {
            
            void root(PluginCommand command) {
                command.on("test1", this::exec);
                command.on("test2", this::exec);
                command.on("test3", this::exec);
            }
            
            void exec(String string) {
                fail("Should not run");
            }
        }
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "", test::root);
        assertThat(completions).containsExactly("test1", "test2", "test3");
    }
    
    @Test
    void testNoTab() {
        class Test {
            void root(PluginCommand command) {
                command.then(this::exec);
            }
            
            void exec(CommandSender sender) {
                fail("Should not run");
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "", test::root);
        
        assertThat(sender.getMessages()).isEmpty();
        assertThat(completions).isEmpty();
    }
    
    @Test
    void testSimpleSubs() {
        class Test {
            void root(PluginCommand command) {
                Consumer<PluginCommand> dontRun = cmd -> fail("should not run");
                command.on("a", dontRun);
                command.on("b", dontRun);
                command.on("c", dontRun);
                command.then(this::exec);
            }
            
            void exec(CommandSender sender) {
                fail("Should not run");
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "", test::root);
        
        assertThat(sender.getMessages()).isEmpty();
        assertThat(completions).containsSequence("a", "b", "c");
    }
    
    @Test
    void testSubMixed() {
        Consumer<PluginCommand> dontRun = cmd -> fail("should not run");
        class Test {
            void root(PluginCommand command) {
                command.on("a", this::cmd);
                command.on("b", this::cmd);
                command.on("c", this::exec);
                command.on("d", this::cmd);
                assertThat(command.state).isEqualTo(State.EXECUTED);
                command.otherwise(() -> dontRun.accept(null));
            }
            
            void cmd(PluginCommand command) {
            }
            
            void exec(CommandSender sender) {
                dontRun.accept(null);
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "a ", test::root);
        
        assertThat(sender.getMessages()).isEmpty();
        assertThat(completions).isEmpty();
    }
    
    @Test
    void testAny() {
        class Test {
            void root(PluginCommand command) {
                Function<String, String> f = ignored -> {
                    fail("should not run");
                    return "";
                };
                command.withAny(f, "a", "b", "c");
                command.withAny(f, "d", "e", "f");
                command.then(this::exec);
            }
            
            void exec(CommandSender sender) {
                fail("Should not run");
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "", test::root);
        
        assertThat(sender.getMessages()).isEmpty();
        assertThat(completions).containsSequence("a", "b", "c", "d", "e", "f");
        
    }
    
    @Test
    void testOnTake1() {
        class Test {
            void root(PluginCommand command) {
                Consumer<PluginCommand> doesNothing = cmd -> {
                };
                command.on("a", doesNothing);
                command.on("b", doesNothing);
                command.on("c", doesNothing);
                command.then(this::exec);
            }
            
            void exec(CommandSender sender) {
                fail("Should not run");
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "", test::root);
        assertThat(completions).containsSequence("a", "b", "c");
        
        completions = tabDiscovery.getCompletions(sender, "a", test::root);
        assertThat(completions).isEmpty();
    }
    
    @Test
    void testInvalidStop() {
        class Test {
            void root(PluginCommand command) {
                Consumer<PluginCommand> doesNothing = cmd -> fail("should not run");
                command.on("a", doesNothing);
                command.on("b", doesNothing);
                command.on("c", doesNothing);
                command.onInvalid(s -> String.format("%s is not valid", s));
                command.then(this::exec);
            }
            
            void exec(CommandSender sender) {
                fail("Should not run");
            }
        }
        
        Test test = new Test();
        var completions = tabDiscovery.getCompletions(sender, "g", test::root);
        assertThat(completions).isEmpty();
    }
}

File: mcspring-api/mcspring-subcommands/src/test/java/in/kyle/mcspring/subcommands/TestTabCompletion.kt
package `in`.kyle.mcspring.subcommands

import `in`.kyle.mcspring.subcommands.TestConsole.run
import `in`.kyle.mcspring.subcommands.plugincommand.api.PluginCommand
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.jupiter.api.Test

internal class TestTabCompletion {

    @Test
    fun testWithString() {
        class Test {
            fun base(command: PluginCommand) {
                command.withString()
                command.withString()
            }
        }

        val command = run("hello world", Test()::base, runExecutors = false)

        assertThat(command.completions).hasSize(2)
        assertThat(command.completions.first().completions).isEmpty()
        assertThat(command.completions.last().completions).isEmpty()
    }

    @Test
    fun testThenNoExecute() {
        class Test {
            fun base(command: PluginCommand) {
                command.then(this::fail)
            }

            fun fail(): Unit = fail("Should not run")
        }

        run("", Test()::base, runExecutors = false)
    }

    @Test
    fun testOnRecurse() {
        class Test {
            fun base(command: PluginCommand) {
                command.on("subcommand-1", this::sub1)
                command.on("other-1", this::fail)
                command.on("other-2", this::fail)
            }

            fun sub1(command: PluginCommand) {
                command.on("subcommand-2", this::fail)
            }

            fun fail(): Unit = fail("Should not run")
        }

        val command = run("subcommand-1 subcommand-2", Test()::base, runExecutors = false)
        assertThat(command.completions).hasSize(1)
        assertThat(command.completions[0].completions).hasSize(3)
        assertThat(command.child).isNotNull()
        assertThat(command.child!!.completions).hasSize(1)
    }

    @Test
    fun testMultiOn() {
        class Test {
            fun base(command: PluginCommand) {
                command.on("a", this::fail)
                command.on("b", this::fail)
                command.on("c", this::fail)
            }

            fun fail(): Unit = fail()
        }

        val command = run("", Test()::base, runExecutors = false)
        assertThat(command.completions).hasSize(1)
        assertThat(command.completions[0].completions).containsExactly("a", "b", "c")
    }

    @Test
    fun testWithAny() {
        class Test {
            fun base(command: PluginCommand) = command.withAny(listOf("a", "b", "c")) { "" }
        }

        val command = run("", Test()::base, runExecutors = false)
        assertThat(command.completions).hasSize(1)
        assertThat(command.completions[0].completions).containsExactly("a", "b", "c")
    }
}
